libcylinder/src/signing/secp256k1/mod.rs

```diff
@@ -52,30 +49,85 @@ impl Default for Secp256k1Context {
 }
 
 impl Context for Secp256k1Context {
-    fn get_algorithm_name(&self) -> &str {
-        "secp256k1"
+    fn new_signer(&self, key: PrivateKey) -> Box<dyn Signer> {
+        Box::new(Secp256k1Signer::new(self.context.clone(), key))
     }
 
-    fn sign(&self, message: &[u8], key: &PrivateKey) -> Result<Signature, Error> {
+    fn new_verifier(&self) -> Box<dyn Verifier> {
+        Box::new(Secp256k1Verifier::new(self.context.clone()))
+    }
+
+    fn new_random_private_key(&self) -> PrivateKey {
+        let mut key = [0u8; secp256k1::constants::SECRET_KEY_SIZE];
+        OsRng.fill_bytes(&mut key);
+        PrivateKey::new(Vec::from(&key[..]))
+    }
+
+    fn get_public_key(&self, private_key: &PrivateKey) -> Result<PublicKey, ContextError> {
+        let sk = secp256k1::key::SecretKey::from_slice(private_key.as_slice())?;
+        Ok(PublicKey::new(
+            secp256k1::key::PublicKey::from_secret_key(&self.context, &sk)
+                .serialize()
+                .to_vec(),
+        ))
+    }
+}
+
+#[derive(Clone)]
+pub struct Secp256k1Signer {
```

Comment:
Does this an verifier impl need to be pub?

Reply:
They don't; I updated them to be private, we can always expose them at a later time if we have a need to.